kore/src/Kore/Equation/Application.hs

```diff
@@ -297,21 +293,28 @@ applyMatchResult equation matchResult@(predicate, substitution) = do
                 }
         _      -> return ()
     let predicate' =
-            Predicate.substitute substitution predicate
-            & Predicate.mapVariables (pure Target.unTarget)
+            Predicate.substitute orientedSubstitution predicate
         equation' =
-            Equation.substitute substitution equation
-            & Equation.mapVariables (pure Target.unTarget)
+            Equation.substitute orientedSubstitution equation
     return (equation', predicate')
   where
-    equationVariables = freeVariables equation & FreeVariables.toList
+    orientedSubstitution = orientSubstitution occursInEquation substitution
+
+    equationVariables = freeVariables equation
+
+    occursInEquation :: (SomeVariableName variable -> Bool)
+    occursInEquation =
+        (`Set.member` equationVariableNames)
+      where
+        equationVariableNames =
+            Set.map variableName (FreeVariables.toSet equationVariables)
```

Comment:
Let's move this to the top-most `where` block. This will prevent it from being newly evaluated each time `occursInEquation` is called. The compiler _should_ do this optimization for us, but I don't like to count on it.

```suggestion

    equationVariableNames =
        Set.map variableName (FreeVariables.toSet equationVariables)
```